activesupport/lib/active_support/json/as_json_encoder.rb

```diff
@@ -0,0 +1,201 @@
+module ActiveSupport
+  module JSON
+    class AsJSONEncoder
+      def self.encode(object, options)
+        new(options).encode object
+      end
+
+      def self.encode_next(object, options)
+        new(options).visit object
+      end
+
+      def self.internal_as_json?(object)
+        object.respond_to?(:as_json) &&
```

Comment:
```suggestion
        !object.respond_to?(:as_json) ||
```

If this method returns false we'll call `as_json`, so probably better to treat something that's undeffed it like an Object.